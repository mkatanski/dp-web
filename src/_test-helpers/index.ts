export * from "./withResponsiveContext";
export * from "./withThemeProvider";
