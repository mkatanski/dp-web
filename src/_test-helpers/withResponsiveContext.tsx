import React from "react";
import { Context as ResponsiveContext } from "react-responsive";
import { Breakpoints } from "config/breakpoints";

export const withMobileSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.mobile }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withTabletSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.tablet }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.desktop }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withLargeDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.largeDesktop }}>
    {Component}
  </ResponsiveContext.Provider>
);
