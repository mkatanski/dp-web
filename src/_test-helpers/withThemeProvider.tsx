import React from "react";
import { ThemeProvider as DefaultThemeProvider } from "lib/theme/ThemeProvider";
import { StyledTheme, Theme } from "styles/theme/_types";
import { createMuiTheme } from "@material-ui/core";

const testTheme: StyledTheme = {
  primary: "primary",
  secondary: "secondary",
  textPrimary: "textPrimary",
  textSecondary: "textSecondary",
  textInverse: "textInverse"
};

const muiTheme = createMuiTheme({
  direction: "ltr"
});

const theme: Theme = Object.assign(testTheme, muiTheme);

export const withThemeProvider = (Component: React.ReactNode) => (
  <DefaultThemeProvider theme={theme}>{Component}</DefaultThemeProvider>
);
