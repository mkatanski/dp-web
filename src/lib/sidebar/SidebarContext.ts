import { createContext } from "react";

export type SidebarStateContextProps = boolean;

export type SidebarToggleContextProps = () => void;

export const SidebarStateContext = createContext<SidebarStateContextProps>(
  false
);

export const SidebarToggleContext = createContext<SidebarToggleContextProps>(
  () => null
);
