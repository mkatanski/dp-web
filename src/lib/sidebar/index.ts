export * from "./SidebarContext";
export * from "./SidebarStateProvider";
