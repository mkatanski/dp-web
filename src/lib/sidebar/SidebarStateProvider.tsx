import React, { useState, useEffect } from "react";
import { useLocation } from "react-router-dom";
import { SidebarStateContext, SidebarToggleContext } from "./SidebarContext";

export type SidebarProviderProps = React.PropsWithChildren<{}>;

export const SidebarStateProvider = ({ children }: SidebarProviderProps) => {
  const [isSidebarOpen, setSidebarOpen] = useState(false);
  const toggleSidebar = () => setSidebarOpen(!isSidebarOpen);
  const { pathname } = useLocation();

  useEffect(() => {
    setSidebarOpen(false);
  }, [pathname]);

  return (
    <SidebarStateContext.Provider value={isSidebarOpen}>
      <SidebarToggleContext.Provider value={toggleSidebar}>
        {children}
      </SidebarToggleContext.Provider>
    </SidebarStateContext.Provider>
  );
};
