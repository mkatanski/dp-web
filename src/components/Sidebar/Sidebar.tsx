import React, { useContext } from "react";
import { useMediaQuery } from "react-responsive";
import styled from "styled-components";
import {
  Drawer as BareDrawer,
  DrawerProps,
  Paper as BarePaper
} from "@material-ui/core";

import { Breakpoints } from "config/breakpoints";
import { SidebarStateContext, SidebarToggleContext } from "lib/sidebar";

enum DrawerPosition {
  top = "top",
  bottom = "bottom",
  left = "left",
  right = "right"
}

const Paper = styled(BarePaper)`
  width: 256px;
  top: 65px;
  height: calc(100% - 65px);
  border: none;
  border-radius: 0;
`;

export type SidebarProps = React.PropsWithChildren<{}>;

export const Sidebar: React.FC<SidebarProps> = ({ children }: SidebarProps) => {
  const isSidebarOpen = useContext(SidebarStateContext);
  const toggleSidebar = useContext(SidebarToggleContext);
  const isLargeDesktop = useMediaQuery({ minWidth: Breakpoints.largeDesktop });
  const commonDrawerProps: Partial<DrawerProps> = {
    anchor: DrawerPosition.left,
    PaperProps: { component: Paper }
  };

  return (
    <>
      {isLargeDesktop ? (
        <BareDrawer {...commonDrawerProps} open variant="persistent">
          {children}
        </BareDrawer>
      ) : (
        <BareDrawer
          {...commonDrawerProps}
          onClose={toggleSidebar}
          open={isSidebarOpen}
          variant="temporary"
        >
          {children}
        </BareDrawer>
      )}
    </>
  );
};
