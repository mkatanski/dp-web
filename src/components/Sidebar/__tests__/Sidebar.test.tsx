import React from "react";
import TestRenderer from "react-test-renderer";
import { SidebarStateContext } from "lib/sidebar";
import "jest-styled-components";

import { withThemeProvider, withLargeDesktopSize } from "_test-helpers";

import { Sidebar } from "../Sidebar";

describe("components", () => {
  describe("Sidebar", () => {
    it("should render Sidebar component", () => {
      const tree = TestRenderer.create(
        withThemeProvider(<Sidebar>children</Sidebar>)
      );
      expect(tree).toMatchSnapshot();
    });

    it("should render Sidebar component with large desktop size", () => {
      const tree = TestRenderer.create(
        withThemeProvider(withLargeDesktopSize(<Sidebar>children</Sidebar>))
      );
      expect(tree).toMatchSnapshot();
    });

    it("should render Sidebar component with open drawer", () => {
      const tree = TestRenderer.create(
        <SidebarStateContext.Provider value={true}>
          <Sidebar />
        </SidebarStateContext.Provider>
      );
      expect(tree).toMatchSnapshot();
    });
  });
});
