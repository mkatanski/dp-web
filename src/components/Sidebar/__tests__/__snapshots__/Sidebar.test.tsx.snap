// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Sidebar should render Sidebar component 1`] = `
<div
  data-props={
    Object {
      "PaperProps": Object {
        "component": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "baseHash": 400283751,
            "componentId": "sc-AxjAm",
            "isStatic": false,
            "rules": Array [
              "
  width: 256px;
  top: 65px;
  height: calc(100% - 65px);
  border: none;
  border-radius: 0;
",
            ],
            "staticRulesId": "",
          },
          "displayName": "Styled(Paper)",
          "foldedComponentIds": Array [],
          "render": [Function],
          "shouldForwardProp": undefined,
          "styledComponentId": "sc-AxjAm",
          "target": [Function],
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      "anchor": "left",
      "onClose": [Function],
      "open": false,
      "variant": "temporary",
    }
  }
  id="MUI.Drawer"
>
  children
</div>
`;

exports[`components Sidebar should render Sidebar component with large desktop size 1`] = `
<div
  data-props={
    Object {
      "PaperProps": Object {
        "component": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "baseHash": 400283751,
            "componentId": "sc-AxjAm",
            "isStatic": false,
            "rules": Array [
              "
  width: 256px;
  top: 65px;
  height: calc(100% - 65px);
  border: none;
  border-radius: 0;
",
            ],
            "staticRulesId": "",
          },
          "displayName": "Styled(Paper)",
          "foldedComponentIds": Array [],
          "render": [Function],
          "shouldForwardProp": undefined,
          "styledComponentId": "sc-AxjAm",
          "target": [Function],
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      "anchor": "left",
      "open": true,
      "variant": "persistent",
    }
  }
  id="MUI.Drawer"
>
  children
</div>
`;

exports[`components Sidebar should render Sidebar component with open drawer 1`] = `
<div
  data-props={
    Object {
      "PaperProps": Object {
        "component": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "baseHash": 400283751,
            "componentId": "sc-AxjAm",
            "isStatic": false,
            "rules": Array [
              "
  width: 256px;
  top: 65px;
  height: calc(100% - 65px);
  border: none;
  border-radius: 0;
",
            ],
            "staticRulesId": "",
          },
          "displayName": "Styled(Paper)",
          "foldedComponentIds": Array [],
          "render": [Function],
          "shouldForwardProp": undefined,
          "styledComponentId": "sc-AxjAm",
          "target": [Function],
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      "anchor": "left",
      "onClose": [Function],
      "open": true,
      "variant": "temporary",
    }
  }
  id="MUI.Drawer"
/>
`;
